[UnifiedPDF] [iOS] Device rotation animation drops multiple frames when viewing unified PDF

whsieh opened the pull request:
#### 9606ca6cea623a50e32abcdaf26780f9066155b4
<pre>
[UnifiedPDF] [iOS] Device rotation animation drops multiple frames when viewing unified PDF
<a href="https://bugs.webkit.org/show_bug.cgi?id=288386">https://bugs.webkit.org/show_bug.cgi?id=288386</a>
<a href="https://rdar.apple.com/137502466">rdar://137502466</a>

Reviewed by Abrar Rahman Protyasha and Tim Horton.

When rotating the device in Safari while viewing a PDF (with unified PDF enabled), `backboardd`
consumes a massive amount of memory (~7-8 GB on an iPad Pro), while attempting to perform a
foreground snapshot of Safari. This massive memory growth can be attributed to CoreAnimation&apos;s
attempts to render shadows behind the page background and container layers, which cause it to try
and encode giant image textures in the Metal-backed context.

For now, work around this by disabling these shadows on iOS family — in the future, we should
restore this codepath, in a way that avoids hitting this performance cliff (likely by moving off of
using platform `CAFilter`s to render shadows).

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFPresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFPresentationController.mm:
(WebKit::PDFPresentationController::makePageContainerLayer):
(WebKit::PDFPresentationController::shouldAddPageBackgroundLayerShadow const):

Canonical link: <a href="https://commits.webkit.org/291001@main">https://commits.webkit.org/291001@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b019dc74b82e7ce72c1c44c3716bebba50515693

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91621 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11151 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/690 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96589 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42293 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11528 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19585 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70352 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27863 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94622 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8806 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83013 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50677 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8570 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41471 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78888 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/605 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98594 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18776 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79378 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19030 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78851 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78586 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19456 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23109 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/464 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11888 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18769 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24040 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18478 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21935 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20239 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFPresentationController.mm:
```diff
@@ -116,27 +116,39 @@
-    addLayerShadow(*pageContainerLayer, containerShadowOffset, containerShadowColor, containerShadowStdDeviation);
+    if (shouldAddPageBackgroundLayerShadow()) {
+        pageContainerLayer->setAnchorPoint({ });
+        addLayerShadow(*pageContainerLayer, containerShadowOffset, containerShadowColor, containerShadowStdDeviation);
```
We only want to limit these calls behind `shouldAddPageBackgroundLayerShadow()` right?

whsieh replied:
...yes, we do. Oops!

whsieh committed:
[UnifiedPDF] [iOS] Device rotation animation drops multiple frames when viewing unified PDF
https://bugs.webkit.org/show_bug.cgi?id=288386
rdar://137502466

Reviewed by Abrar Rahman Protyasha and Tim Horton.

When rotating the device in Safari while viewing a PDF (with unified PDF enabled), `backboardd`
consumes a massive amount of memory (~7-8 GB on an iPad Pro), while attempting to perform a
foreground snapshot of Safari. This massive memory growth can be attributed to CoreAnimation's
attempts to render shadows behind the page background and container layers, which cause it to try
and encode giant image textures in the Metal-backed context.

For now, work around this by disabling these shadows on iOS family — in the future, we should
restore this codepath, in a way that avoids hitting this performance cliff (likely by moving off of
using platform `CAFilter`s to render shadows).

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFPresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFPresentationController.mm:
(WebKit::PDFPresentationController::makePageContainerLayer):
(WebKit::PDFPresentationController::shouldAddPageBackgroundLayerShadow const):

Canonical link: https://commits.webkit.org/291001@main
